File: cpp/33.cpp
class Solution {
public:
    int search(int A[], int n, int target) {
       return f(A,0,n-1,target); 
    }
    int f(int A[],int l,int r,int target){
    	if(l>r)return -1;
    	int m=l+(r-l)/2;
    	if(A[m]==target) return m;
    	if(A[m]>=A[l]){
    		if(target>=A[l]&&target<A[m])
    			return f(A,l,m-1,target);
    		else
    			return f(A,m+1,r,target);
    	}
    	if(A[r]>=A[m]){
    		if(target>A[m]&&target<=A[r])
    			return f(A,m+1,r,target);
    		else
    			return f(A,l,m-1,target);
    	}
    	return -1;
    }
};

File: js/33.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
  let l = 0, r = nums.length - 1
  while(l < r) {
    let m = (l + r) >> 1 
    if(nums[m] == target) return m 
    if(nums[l] <= nums[m]) {
      if(target >= nums[l] && target < nums[m]) {
        r = m - 1
      } else {
        l = m + 1
      }
    } else {
      if(target > nums[m] && target <= nums[r]) {
        l = m + 1
      } else {
        r = m - 1
      }
    }
  }
  return nums[l] == target ? l : -1
};